compiler/rustc_const_eval/src/const_eval/panic.rs

```diff
@@ -0,0 +1,262 @@
+use rustc_hir::def_id::DefId;
+use rustc_middle::ty::{self, layout::LayoutOf, subst::Subst};
+use rustc_span::sym;
+use std::cell::Cell;
+use std::fmt::{self, Debug, Formatter};
+
+use crate::interpret::{FnVal, InterpCx, InterpErrorInfo, InterpResult, OpTy};
+
+use super::CompileTimeInterpreter;
+
+struct Arg<'mir, 'tcx, 'err> {
+    cx: &'err InterpCx<'mir, 'tcx, CompileTimeInterpreter<'mir, 'tcx>>,
+    arg: OpTy<'tcx>,
+    fmt_trait: DefId,
+    err: &'err Cell<Option<InterpErrorInfo<'tcx>>>,
+}
+
+impl Debug for Arg<'_, '_, '_> {
+    fn fmt(&self, f: &mut Formatter<'_>) -> fmt::Result {
+        match self.cx.fmt_arg(self.arg, self.fmt_trait, f) {
+            Ok(_) => Ok(()),
+            Err(e) => {
+                self.err.set(Some(e));
+                Err(fmt::Error)
+            }
+        }
+    }
+}
+
+impl<'mir, 'tcx> InterpCx<'mir, 'tcx, CompileTimeInterpreter<'mir, 'tcx>> {
+    fn fmt_arg(
+        &self,
+        arg: OpTy<'tcx>,
+        fmt_trait: DefId,
+        f: &mut Formatter<'_>,
+    ) -> InterpResult<'tcx> {
+        let fmt_trait_sym = self.tcx.item_name(fmt_trait);
+        let fmt_trait_name = fmt_trait_sym.as_str();
+
+        macro_rules! dispatch_fmt {
+            ($e: expr, $($t: ident)|*) => {
+                let _ = match &*fmt_trait_name {
+                    $(stringify!($t) => fmt::$t::fmt($e, f),)*
+                    _ => Debug::fmt($e, f),
+                };
+            }
+        }
+
+        match arg.layout.ty.kind() {
+            ty::Bool => {
+                let v = self.read_scalar(&arg)?.to_bool()?;
+                dispatch_fmt!(&v, Display);
+            }
+            ty::Char => {
+                let v = self.read_scalar(&arg)?.to_char()?;
+                dispatch_fmt!(&v, Display);
+            }
+            ty::Int(int_ty) => {
+                let v = self.read_scalar(&arg)?.check_init()?;
+                let v = match int_ty {
+                    ty::IntTy::I8 => v.to_i8()?.into(),
+                    ty::IntTy::I16 => v.to_i16()?.into(),
+                    ty::IntTy::I32 => v.to_i32()?.into(),
+                    ty::IntTy::I64 => v.to_i64()?.into(),
+                    ty::IntTy::I128 => v.to_i128()?,
+                    ty::IntTy::Isize => v.to_machine_isize(self)?.into(),
+                };
+                dispatch_fmt!(
+                    &v,
+                    Display | Binary | Octal | LowerHex | UpperHex | LowerExp | UpperExp
+                );
```

Comment:
This makes some assumptions about how integers are formatted, namely that using a larger integer type will not make a difference. This assumption should at least be explicitly documented in a comment here.

Also, do these `LowerHex` etc things her mean that hex output is supported? I don't see the tests covering that possibility.